burstphoto/align.metal

```diff
@@ -416,3 +740,992 @@ kernel void compute_merge_weight(texture2d<float, access::read> texture_diff [[t
+    mismatch_norm = clamp(mismatch_norm, 0.0f, 1.0f);
+    
+    mismatch_texture.write(mismatch_norm, gid);
+}
+
+
+// hot pixel correction based on the idea that hot pixels appear at the same pixel location in all images
+kernel void correct_hotpixels(texture2d<float, access::read> average_texture [[texture(0)]],
+                              texture2d<float, access::read> in_texture [[texture(1)]],
+                              texture2d<float, access::write> out_texture [[texture(2)]],
+                              constant float* mean_texture_buffer [[buffer(0)]],
+                              uint2 gid [[thread_position_in_grid]]) {
+       
+    int const x = gid.x+2;
+    int const y = gid.y+2;
+    
+    // load args
+    float mean_texture = 0.0f;
+    
+    // extract color channel-dependent mean value of the average texture of all images
+    if (x%2 == 0 & y%2 == 0) {
+        mean_texture = mean_texture_buffer[0];
+    }
+    if (x%2 == 1 & y%2 == 0) {
+        mean_texture = mean_texture_buffer[1];
+    }
+    if (x%2 == 0 & y%2 == 1) {
+        mean_texture = mean_texture_buffer[2];
+    }
+    if (x%2 == 1 & y%2 == 1) {
+        mean_texture = mean_texture_buffer[3];
+    }
+        
+    // calculate weighted sum of 8 pixels surrounding the potential hot pixel based on the average texture
+    float sum =   average_texture.read(uint2(x-2, y-2)).r;
+    sum      +=   average_texture.read(uint2(x+2, y-2)).r;
+    sum      +=   average_texture.read(uint2(x-2, y+2)).r;
+    sum      +=   average_texture.read(uint2(x+2, y+2)).r;
+    sum      += 2*average_texture.read(uint2(x-2, y+0)).r;
+    sum      += 2*average_texture.read(uint2(x+2, y+0)).r;
+    sum      += 2*average_texture.read(uint2(x+0, y-2)).r;
+    sum      += 2*average_texture.read(uint2(x+0, y+2)).r;
+    
+    // extract value of potential hot pixel from the average texture and divide by sum of surrounding pixels
+    float const pixel_value = average_texture.read(uint2(x, y)).r;
+    float const pixel_ratio = pixel_value/sum;
+    
+    // if hot pixel is detected
+    if (pixel_ratio >= 0.15f & pixel_value >= 2.0f*mean_texture) {
+        
+        // calculate mean value of 4 surrounding values
+        float sum2 = in_texture.read(uint2(x-2, y+0)).r;
+        sum2      += in_texture.read(uint2(x+2, y+0)).r;
+        sum2      += in_texture.read(uint2(x+0, y-2)).r;
+        sum2      += in_texture.read(uint2(x+0, y+2)).r;
+        
+        // calculate weight for blending to have a smooth transition for not so obvious hot pixels
+        float const weight = 10.0f*min(pixel_ratio-0.15f, 0.10f);
+        
+        // blend values and replace hot pixel value
+        out_texture.write(weight*0.25f*sum2 + (1.0f-weight)*in_texture.read(uint2(x, y)).r, uint2(x, y));
+    }
+}
+
+
+// highly-optimized fast Fourier transform applied to each color channel independently
+kernel void forward_fft(texture2d<float, access::read> in_texture [[texture(0)]],
```

Comment:
I think this should be named something like `forward_fast_dft`, I think that `forward_fft` is a confusing name since this is a *fast discrete Fourier transform* and not a *fast Fourier transform*.

Reply:
The general term for this kind of algorithm, which calculates a discrete Fourier transform with an order of N*log(N), is "Fast Fourier Transform". Therefore I believe the name should stay as it is.